File: Array/283_MoveZeroes.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com


class Solution(object):
    def moveZeroes(self, nums):
        # Shift non-zero values as far forward as possible.
        index = 0
        for n in nums:
            if n != 0:
                nums[index] = n
                index += 1

        # Fill remaining space with zeros
        for i in range(index, len(nums)):
            nums[i] = 0

        return

"""
[]
[1]
[0]
[0,0,0]
[0,1,0,3,12]
[7,6,5,4,0,4,0,5,6,0,7,0,0]
"""

File: Array/283_MoveZeroes.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-28 10:57:41
 */

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        // Every time put the non-zero number to the right position.
        int non_zero = 0, index = 0;
        for(; index<nums.size(); index++){
            if(nums[index] != 0){
                if(index != non_zero){
                    swap(nums[index], nums[non_zero]);
                }
                non_zero += 1;
            }
        }
    }
};

/*
[]
[1]
[0]
[0,0,0]
[0,1,0,3,12]
[7,6,5,4,0,4,0,5,6,0,7,0,0]
*/
